Source/Dafny/Resolver.cs

```diff
@@ -3488,11 +3503,19 @@ public class ModuleBindings {
           } else if (member is Function) {
             var f = (Function)member;
             ResolveParameterDefaultValues_Pass1(f.Formals, f);
-            if (!f.IsGhost && f.Body != null) {
-              CheckIsCompilable(f.Body);
-            }
-            if (f.Body != null) {
-              DetermineTailRecursion(f);
+            if (f.ByMethodBody == null) {
+              if (!f.IsGhost && f.Body != null) {
+                CheckIsCompilable(f.Body, f);
+              }
+              if (f.Body != null) {
+                DetermineTailRecursion(f);
+              }
+            } else {
+              var m = f.ByMethodDecl;
+              Contract.Assert(m != null && !m.IsGhost);
+              ComputeGhostInterest(m.Body, m.IsGhost, m);
+              CheckExpression(m.Body, this, m);
```

Comment:
I'm a bit confused about the `CheckExpression` call. Both it and `ComputeGhostInterest` seem to relate to the same thing. Why do we need two calls?

Reply:
I don't remember the reason for this. As it stands, `ComputeGhostInterest` sets the `.IsGhost` fields for most statements, but `CheckExpression` does that for the statement part of `StmtExpr`s. Also, `CheckExpression` enforces that statements inside expressions are legal, which I _think_ requires `ComputeGhostInterest` to have done its work first. It's entirely possible that these could either be compiled or refactored into more orthogonal pieces, but that lies outside the scope of this PR. For now, this code does the same thing for a by-method as it does for an ordinary method.